File: services/routing-service/consumer.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/segmentio/kafka-go"
)

const maxProcessingAttempts = 3

type RetryBackoff interface {
	Wait(attempt int)
}

type ExponentialBackoff struct{}

func (ExponentialBackoff) Wait(attempt int) {
	delay := time.Second * time.Duration(1<<(attempt-1))
	time.Sleep(delay)
}

type PermanentProcessingError struct {
	Err error
}

func (e *PermanentProcessingError) Error() string {
	return e.Err.Error()
}

func (e *PermanentProcessingError) Unwrap() error {
	return e.Err
}

func shouldRetry(err error) bool {
	var permanentErr *PermanentProcessingError
	return !errors.As(err, &permanentErr)
}

func processShipmentWithRetry(
	graph Graph,
	event ShipmentCreatedEvent,
	store ProcessedEventStore,
	maxAttempts int,
	backoff RetryBackoff,
) error {
	var lastErr error

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		if err := processShipment(
			graph,
			event,
			store,
		); err != nil {
			if !shouldRetry(err) {
				return err
			}

			if attempt < maxAttempts {
				backoff.Wait(attempt)
			}

			lastErr = err
			continue
		}

		return nil
	}

	return lastErr
}

// handleKafkaMessage processes a single Kafka message.
// Transient failures use bounded retries with exponential backoff.
// Permanent failures skip retries. Permanent failures and exhausted retries
// are published to the DLQ before the original Kafka offset is committed.
func handleKafkaMessage(
	ctx context.Context,
	message kafka.Message,
	graph Graph,
	store ProcessedEventStore,
	committer MessageCommitter,
	deadLetterPublisher DeadLetterPublisher,
	backoff RetryBackoff,
) error {
	var event ShipmentCreatedEvent

	if err := json.Unmarshal(message.Value, &event); err != nil {
		if err := deadLetterPublisher.PublishDeadLetter(
			ctx,
			message,
			"invalid JSON",
		); err != nil {
			return fmt.Errorf("publish dead-letter message: %w", err)
		}

		return committer.CommitMessages(ctx, message)
	}

	// Multiple event types share the topic; Routing Service only handles ShipmentCreated.
	if event.EventType != "ShipmentCreated" {
		return committer.CommitMessages(ctx, message)
	}

	// Processing must reach a terminal outcome before the Kafka offset is committed:
	// either successful processing or successful publication to the DLQ.
	// If DLQ publication fails, the offset remains uncommitted so Kafka can redeliver.
	if err := processShipmentWithRetry(
		graph,
		event,
		store,
		maxProcessingAttempts,
		backoff,
	); err != nil {
		reason := "processing retries exhausted"

		var permanentErr *PermanentProcessingError
		if errors.As(err, &permanentErr) {
			reason = "permanent processing failure"
		}

		if dlqErr := deadLetterPublisher.PublishDeadLetter(
			ctx,
			message,
			reason,
		); dlqErr != nil {
			return fmt.Errorf("publish dead-letter message: %w", dlqErr)
		}

		return committer.CommitMessages(ctx, message)
	}

	return committer.CommitMessages(ctx, message)
}

File: services/prediction-service/src/prediction_service/consumer.py
import json
import time

from pydantic import ValidationError

from prediction_service.models import RouteCalculatedEvent
from prediction_service.prediction import handle_route_calculated
from prediction_service.publisher import DeadLetterPublisher
from prediction_service.store import PostgresPredictionStore

MAX_PROCESSING_ATTEMPTS = 3


class RetryBackoff:
    def wait(self, attempt: int) -> None:
        time.sleep(2 ** (attempt - 1))


def handle_kafka_message(
    message,
    store: PostgresPredictionStore,
    consumer,
    dead_letter_publisher: DeadLetterPublisher,
    backoff: RetryBackoff,
) -> None:
    try:
        data = json.loads(message.value().decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        dead_letter_publisher.publish_dead_letter(
            message,
            "invalid JSON",
        )
        consumer.commit(
            message=message,
            asynchronous=False,
        )
        return

    # Multiple event types share the topic; this service only handles RouteCalculated.
    if data.get("event_type") != "RouteCalculated":
        consumer.commit(message=message, asynchronous=False)
        return

    try:
        event = RouteCalculatedEvent.model_validate(data)
    except ValidationError:
        dead_letter_publisher.publish_dead_letter(
            message,
            "invalid RouteCalculated event",
        )
        consumer.commit(
            message=message,
            asynchronous=False,
        )
        return

    for attempt in range(1, MAX_PROCESSING_ATTEMPTS + 1):
        try:
            eta_event = handle_route_calculated(
                event,
                store,
            )
            break
        except Exception:
            if attempt == MAX_PROCESSING_ATTEMPTS:
                dead_letter_publisher.publish_dead_letter(
                    message,
                    "processing retries exhausted",
                )
                consumer.commit(
                    message=message,
                    asynchronous=False,
                )
                return

            backoff.wait(attempt)

    # Commit the Kafka offset only after processing has completed successfully.
    # If the commit fails, persistent event-id idempotency makes redelivery safe.
    consumer.commit(message=message, asynchronous=False)

    if eta_event is None:
        return

    print(
        f"ETA predicted and queued for shipment {eta_event.shipment_id}: "
        f"{eta_event.payload.estimated_travel_minutes} minutes"
    )
